igvm/src/c_api.rs

```diff
@@ -0,0 +1,366 @@
+    }
+    IgvmApiError::IGVMAPI_OK
+}
+
+/// Parse a binary array containing an IGVM file.
+/// The contents of the file are validated and, if valid, a handle is returned
+/// to represent the file. This handle must be freed with a call to
+/// igvm_free().
+///
+/// # Safety
+/// This function is exported and designed to be called as a C API. As such
+/// it works with raw pointers and is inherently unsafe due to this.
+#[no_mangle]
+pub unsafe extern "C" fn igvm_new_from_binary(
+    data: *const u8,
+    len: u32,
+    igvm: *mut *mut IgvmFile,
+) -> IgvmApiError {
+    // Safety: data array must be at least as large as len
+    let file_data = unsafe { std::slice::from_raw_parts(data, len as usize) };
+    let result = IgvmFile::new_from_binary(file_data, None);
+
+    match result {
+        Ok(file) => {
+            // Safety: Dereferences the C-provided pointer
+            unsafe {
+                *igvm = Box::into_raw(Box::new(file));
+            }
+            IgvmApiError::IGVMAPI_OK
+        }
+        Err(e) => translate_error(e),
+    }
+}
+
+/// Free a handle that was created with this library.
+///
+/// # Safety
+/// This function is exported and designed to be called as a C API. As such
+/// it works with raw pointers and is inherently unsafe due to this.
+#[no_mangle]
+pub unsafe extern "C" fn igvm_free(igvm: *mut IgvmFile) {
+    // Safety: Using structure provided as a raw pointer from C
+    unsafe {
+        drop(Box::from_raw(igvm));
+    };
+}
+
+/// Get the count of headers for a particular section in the parsed IGVM
+/// file.
+///
+/// # Safety
+/// This function is exported and designed to be called as a C API. As such
+/// it works with raw pointers and is inherently unsafe due to this.
+#[no_mangle]
+pub unsafe extern "C" fn igvm_header_count(igvm: *mut IgvmFile, section: IgvmHeaderSection) -> u32 {
+    // Safety: Using pointer from C
+    unsafe {
+        let len = match section {
+            IgvmHeaderSection::HEADER_SECTION_PLATFORM => (*igvm).platform_headers.len(),
+            IgvmHeaderSection::HEADER_SECTION_INITIALIZATION => {
+                (*igvm).initialization_headers.len()
+            }
+            IgvmHeaderSection::HEADER_SECTION_DIRECTIVE => (*igvm).directive_headers.len(),
+            _ => 0,
+        };
+        len as u32
+    }
+}
+
+/// Get the header type for the entry with the given index for a particular
+/// section in the parsed IGVM file.
+///
+/// This function can be called to determine which IGVM_VHS_* structure will
+/// be returned by a call to igvm_get_header().
+///
+/// # Safety
+/// This function is exported and designed to be called as a C API. As such
+/// it works with raw pointers and is inherently unsafe due to this.
+#[no_mangle]
+pub unsafe extern "C" fn igvm_get_header_type(
+    igvm: *mut IgvmFile,
+    section: IgvmHeaderSection,
+    index: u32,
+    typ: *mut IgvmVariableHeaderType,
+) -> IgvmApiError {
+    // Safety: Using pointer from C
+    unsafe {
```

Comment:
We should limit the scope of the unsafe block here as much as we can. Maybe use some local variables to create rust references, then assign the type at the very end rather than needing multiple smaller unsafe blocks throughout.

Reply:
I've managed to remove many of the unsafe regions and reworked the remainder to the smallest scope possible.